### PR TITLE
[Lifter] Complex type bugfixes

### PR DESCRIPTION
**Pull request by blattm**
The following issues related to lifting types were found:
- [x] 1) Infinite recursion when lifting complex types
- [x] 2) `RuntimeError: Unknown struct type ClassStructureType`
- [x] 3) `Enum.get_name_by_value` crashes if value is not part of enum
- [x] 4) `AssignmentHandler.lift_set_field` crashes for different reasons
- [x] 5) `UnaryOperationHandler._lift_load_struct` can't handle offsets other than members's.
- [x] 6) Infinite recursions in generation of declaration. 

Remarks:
1) The preexisting strategy of preventing infinite recursion was not touched. It is unclear if it is still needed. Maybe only a single strategy should be used?
2) Right now Structs and Classes behave identically. `_BaseStruct` is considered and implementation detail, but this could be changed.
3) Maybe such enums could be represented like `Enum.Flag1 | Enum.Flag2 | ... | const` in the future.
4) This seemed to happen for `NamedTypeReferenceType`s which did not correspond to structs or classes. Should be fixed now, but I'm unsure about edge cases.
5) When the problem was "fixed" using fake members like `.__offset_0x20`, later stages failed because of missing definitions. It is unclear if this means that the "fix" is insufficient or if these problems are unrelated. EDIT: completely fixed by 5e01dd7. `OperationType.member_access` needs to be treated like `OperationType.dereference`.
6) The problem seems to be name collisions of (anonymous and nested) complex types with name `''`. EDIT: Fixed by comparing with and id instead of a name during lifting. Empty names are replaced by "`__anonymous_struct`", "`__anonymous_class`" or "`__anonymous_union`" . Name collisions of distinct types are prevented by appending "`__2`", "`__3`", etc. to names. Therefore all later stages can use names as identifiers.

### CODE REVIEW

File: decompiler/frontend/binaryninja/handlers/types.py [hunk 1]
```diff
@@ -75,30 +75,46 @@ def lift_named_type_reference_type(self, custom: NamedTypeReferenceType, **kwarg
...
-            type_name = name if name else self._get_data_type_name(struct, keyword="union")
-            lifted_struct = Union_(struct.width * self.BYTE_SIZE, type_name, [])
+            keyword, type, members = "union", Union_, []
+        elif struct.type == StructureVariant.ClassStructureType:
+            keyword, type, members = "class", Class, {}
         else:
             raise RuntimeError(f"Unknown struct type {struct.type.name}")
+
+        type_name = name if name else self._get_data_type_name(struct, keyword=keyword)
+        if type_name.strip() == "":
```
Comment (mari-mari) [inline decompiler/frontend/binaryninja/handlers/types.py:110]: Lines 110-112 looks almost identical to 80-82. Do you think it makes sense to move them to the separate method?

File: decompiler/frontend/binaryninja/handlers/types.py [hunk 2]
```diff
@@ -117,7 +133,8 @@ def lift_struct_member(self, member: StructureMember, parent_struct_name: str =
         else:
             # if member is an embedded struct/union, the name is already available
             member_type = self._lifter.lift(member.type, name=member.name)
-        return ComplexTypeMember(0, name=member.name, offset=member.offset, type=member_type)
+            # This is still wrong for Pointers...
```
Comment (mari-mari) [inline decompiler/frontend/binaryninja/handlers/types.py:136]: If it is still wrong, then a separate issue should be opened :sweat_smile:

File: decompiler/frontend/binaryninja/handlers/types.py (hunk 1 above)
Comment (blattm) [reply decompiler/frontend/binaryninja/handlers/types.py:110]: Yes, sounds good! I could add this functionality into the existing `_get_data_type_name`.

File: decompiler/structures/pseudo/complextypes.py [hunk 3]
```diff
@@ -117,19 +132,37 @@ def __str__(self) -> str:
         return self.name
 
 
+class UniqueNameProvider:
```
[the remark this answers is not in the document]
Comment (blattm) [reply decompiler/structures/pseudo/complextypes.py:135]: fixed by [09919b0](https://github.com/fkie-cad/dewolf/pull/353/commits/09919b03c4a7fbfe09f4dd734147021e153b1590)

File: decompiler/frontend/binaryninja/handlers/types.py (hunk 1 above)
Comment (blattm) [reply decompiler/frontend/binaryninja/handlers/types.py]: fixed by [9036c07](https://github.com/fkie-cad/dewolf/pull/353/commits/9036c07e9af6964b8209c31ab197407203d7a2a3)

File: decompiler/frontend/binaryninja/handlers/types.py (hunk 2 above)
Comment (blattm) [reply decompiler/frontend/binaryninja/handlers/types.py]: I debugged it again and everything is actually working as intended. I removed the misleading comment in [e402d24](https://github.com/fkie-cad/dewolf/pull/353/commits/e402d24a706373da71d4bf8fb5c2fdeb2585a9c5).

File: decompiler/structures/pseudo/complextypes.py [hunk 4]
```diff
@@ -117,19 +132,46 @@ def __str__(self) -> str:
...
+    """ The purpose of this class is to provide unique names for types, as duplicate names can potentially be encountered in the lifting stage (especially anonymous structs, etc.)
+    This class keeps track of all the names already used. If duplicates are found, they are renamed by appending suffixes with incrementing numbers.
+    E.g. `classname`, `classname__2`, `classname__3`, ...
+    """
+
+    def __init__(self):
+        self._name_to_count: Dict[str, int] = {}
+
+    def get_unique_name(self, name: str):
```
Comment (mari-mari) [inline decompiler/structures/pseudo/complextypes.py:144]: return type is missing :sweat_smile:

**Pull request merged**